sdk/search/Azure.AI.Search/src/SearchIndexClient.cs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.AI.Search
+{
+    /// <summary>
+    /// Azure Cognitive Search client that can be used to query an index and
+    /// upload, merge, or delete documents.
+    /// </summary>
+    public class SearchIndexClient
+    {
+        /// <summary>
+        /// Gets the URI endpoint of the Search Service.  This is likely
+        /// to be similar to "https://{search_service}.search.windows.net".
+        /// </summary>
+        public virtual Uri Endpoint { get; }
+
+        /// <summary>
+        /// The name of the Search Service, lazily obtained from the
+        /// <see cref="Endpoint"/>.
+        /// </summary>
+        private string _serviceName = null;
+
+        /// <summary>
+        /// Gets the name of the Search Service.
+        /// </summary>
+        public virtual string ServiceName =>
+            _serviceName ??= Endpoint.GetSearchServiceName();
+
+        /// <summary>
+        /// Gets the name of the Search Index.
+        /// </summary>
+        public virtual string IndexName { get; }
+
+        /// <summary>
+        /// Gets the authenticatd <see cref="HttpPipeline"/> used for sending
+        /// requests to the Search Service.
+        /// </summary>
+        private HttpPipeline Pipeline { get; }
```

Comment:
Nit: would make read-only fields `readonly` to help make sure we don't accidentally overwrite.

Reply:
We'd have to create separate backing fields and I'm not sure the visual clutter is worth it.  We started that way with Storage because Jeff felt strongly that immutable things should be enforced as such but eventually worked our way to auto-properties once we had about 10 properties per client.  I could be talked into making the change if folks feel strongly.

Reply:
I don't feel strongly, no. But why not just use a field instead of the property? The compiler or JIT'er is likely to optimize out the method call and just use the backing field anyway. Again, though, don't care that much (hence the "nit"). Just enough to type this.